portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -73,5 +90,44 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     private String getParameter(HttpServletRequest request, String name, String defaultValue) {
         return Optional.ofNullable(request.getParameter(name)).orElse(defaultValue);
-  }
+    }
+
+
+    /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+    private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+        BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+        Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+        List<BlobKey> blobKeys = blobs.get("image");
+
+        // User submitted form without selecting a file, so we can't get a URL. (dev server)
+        if (blobKeys == null || blobKeys.isEmpty()) {
+        return null;
+        }
+
+        // Our form only contains a single file input, so get the first index.
+        BlobKey blobKey = blobKeys.get(0);
+
+        // User submitted form without selecting a file, so we can't get a URL. (live server)
+        BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+        if (blobInfo.getSize() == 0) {
+        blobstoreService.delete(blobKey);
```

Comment:
Is there anything to delete here? If user didn't upload a file?

Reply:
I'm guessing this is part of the tutorial. From the comments about this section being for the live server, I'll hazard that the blobs.get returns a blobKey for any image tag, even if it's empty, and in the liveserver these keys need to be deleted. The app engine docs for this aren't very helpful in explaining the flow.

Reply:
The database doesn't store the imageurl if no image was entered. I ended up checking if the value was null in the JavaScript that way it would't create a new image tag everytime even if the user did not submit an image.